rfcs/0041-new-transaction-model.md

```diff
@@ -0,0 +1,183 @@
+# Redesign of the transaction execution model
+
+## Summary
+
+This proposes refactoring the transaction execution model, primarily to make it easy to implement Snapps transactions involving an arbitrary number of parties without having circuits that verify corresponding numbers of proofs. I.e., this model will make it possible to e.g., have a transaction involving 10 snapp accounts while only having a base transaction SNARK circuit that verifies a single proof.
+
+## Introduction
+
+Currently, the transaction execution part of the protocol can be thought of as running a state machine where the state consists of various things, roughly
+
+- current staged ledger
+- a signed amount (the fee excess)
+- the pending coinbase stack
+- the next available token ID
+ 
+We propose extending this state to include an optional field called
+`current_transaction_local_state` 
+consisting of
+- a hash, called `id`
+- a signed amount, called `excess`
+- an optional token ID, called `current_token_id`
+- a non-empty stack of "parties" (described below), called `remaining_parties`
+
+## Transactions
+  
+Under this approach, a transaction would semantically be a sequence of "parties". A "party" is a sequence of tuples `(authorization, predicate, update)` where
+
+- An authorization is one of
+	- a proof
+	- a signature
+	- nothing
+- A predicate is an encoding of a function `Account.t -> bool`
+- An update is an encoding of a function `Account.t -> Account.t`
+
+For example, a normal payment transaction from an account at nonce `nonce` for amount `amount` with fee `fee` would be (in pseduocaml) the sequence
+
+```ocaml
+[ { authorization= Signature ...
+  ; predicate= (fun a -> a.nonce == nonce)
+  ; update= (fun a -> {a with balance= a.balance - (amount + fee)})
+  }
+; { authorization= Nothing
+  ; predicate= (fun _ -> true)
+  ; update= (fun a -> {a with balance= a.balance + amount})
+  }
+]
+```
+
+A token swap trading `n` of token `A` from `sender_A` for `m` of token `B` from `sender_B`, plus a fee payment of `fee` from `fee_payer` would look like
+```ocaml
+[ { authorization= Signature ...
```

Comment:
Could it be useful to put all the signatures in one place so that they could be replaced with an aggregate signature at a later stage, or possibly aggregate signatures from the start?